src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+
+    T0 = const%water_freeze_temp
+    do i_part = 1, aero_state_n_part(aero_state)
+       a_INAS = -0.517 
+       b_INAS = 8.934
+       aerosol_diameter = aero_particle_dry_diameter(aero_state%apa%particle(i_part), aero_data)
+       S = const%pi * aerosol_diameter **2
+       p = pmc_random()
+       temp = (log(1 - p) + exp(-S * exp(-a_INAS * T0 + b_INAS))) / (-S)
+       aero_state%apa%particle(i_part)%imf_temperature = T0 + (log(temp) &
+          - b_INAS) / a_INAS
+    end do
+  end subroutine singular_initialize
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Simulation for singular scheme, deciding whether to freeze for each
+  !> particle. Run in each time step.
+  subroutine immersion_freezing_singular(aero_state, aero_data, env_state_initial, env_state_final)
+    implicit none
+    !> Aerosol state.
+    type(aero_state_t), intent(inout) :: aero_state
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+    !> Environment state at the start of the timestep.
+    type(env_state_t), intent(in) :: env_state_initial
+    !> Environment state at the end of the timestep. The rel_humid
+    !> value will be ignored and overwritten with a new value.
+    type(env_state_t), intent(inout) :: env_state_final
+
+    real(kind=dp), allocatable :: H2O_masses(:), total_masses(:), &
+        H2O_frac(:)
+    integer :: i_part, i_bin, i_class, n_bins, n_class
+    
+    allocate(total_masses(aero_state_n_part(aero_state)))
+    allocate(H2O_masses(aero_state_n_part(aero_state)))
+    allocate(H2O_frac(aero_state_n_part(aero_state)))
+
+    total_masses = aero_state_masses(aero_state, aero_data)
+    H2O_masses = aero_state_masses(aero_state, aero_data, include=(/"H2O"/))
+    H2O_frac = H2O_masses / total_masses
+    do i_part = 1, aero_state_n_part(aero_state)
+       if (aero_state%apa%particle(i_part)%frozen) then
+          cycle
+       end if
+       if (H2O_frac(i_part) < 1e-2) then
+          cycle
+       end if
+       if (env_state_final%temp .le. &
+            aero_state%apa%particle(i_part)%imf_temperature) then
+          aero_state%apa%particle(i_part)%frozen = .TRUE.
+          aero_state%apa%particle(i_part)%den_ice = &
+               const%reference_ice_density
+          aero_state%apa%particle(i_part)%ice_shape_phi = 1d0
+       end if
+    end do
+
+    deallocate(total_masses)
+    deallocate(H2O_masses)
+    deallocate(H2O_frac)
+
+  end subroutine immersion_freezing_singular
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Simulation for time-dependent scheme (e.g., ABIFM, constant rate),
+  !> deciding whether to freeze for each particle. Run in each time step.   
+  !> This subroutine applys the binned-tau leaping algorithm for speeding up.
+  subroutine immersion_freezing_time_dependent(aero_state, aero_data, env_state_initial, &
+       env_state_final, del_t, immersion_freezing_scheme_type, freezing_rate)
```

Comment:
`env_state_initial` is passed here and never used (and if this is correct, `env_state_final` can just become `env_state`).